zarr/array.py

```diff
@@ -0,0 +1,428 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import, print_function, division
+from functools import reduce  # TODO PY2 compatibility
+import operator
+import itertools
+import multiprocessing
+
+
+import numpy as np
+
+
+from zarr import blosc
+from zarr.util import is_total_slice, normalize_array_selection, \
+    get_chunk_range, human_readable_size
+
+
+# import sys
+#
+# def log(*msg):
+#     print(*msg, file=sys.stderr)
+#     sys.stderr.flush()
+
+
+_blosc_use_context = False
+
+
+def set_blosc_options(use_context=False, nthreads=None):
+    """Set options for how the blosc compressor is used.
+
+    Parameters
+    ----------
+    use_context : bool, optional
+        If False, blosc will be used in non-contextual mode, which is best
+        when using zarr in a single-threaded environment because it allows
+        blosc to use multiple threads internally. If True, blosc will be used
+        in contextual mode, which is better when using zarr in a
+        multi-threaded environment like dask.array because it avoids the blosc
+        global lock and so multiple blosc operations can be running
+        concurrently.
+    nthreads : int, optional
+        Number of internal threads to use when running blosc in non-contextual
+        mode.
+
+    """
+    global _blosc_use_context
+    _blosc_use_context = use_context
+    if not use_context:
+        if nthreads is None:
+            # diminishing returns beyond 4 threads?
+            nthreads = min(4, multiprocessing.cpu_count())
+        blosc.set_nthreads(nthreads)
+
+
+_repr_shuffle = [
+    '0 (NOSHUFFLE)',
+    '1 (BYTESHUFFLE)',
+    '2 (BITSHUFFLE)',
+]
+
+
+class Array(object):
+
+    def __init__(self, store):
+        """Instantiate an array.
+
+        Parameters
+        ----------
+        store : zarr.store.base.ArrayStore
+            Array store.
+
+        """
+
+        self._store = store
+
+    @property
+    def store(self):
+        return self._store
+
+    @property
+    def shape(self):
+        return self._store.meta['shape']
```

Comment:
We may want to cache the metadata on the object itself for repeated reads.  Some `MutableMapping`s will have a non-trivial latency.